Bug: coldfront.core.project.test_views.ProjectListViewTest.test_project_list_search is Flaky

### What happened?

A test in the project application is very flaky. Most of the time it will not fail but every once in awhile it will. 

The most likely reason could think of for this test failing is because the test was mistakenly written believing the Factories always create unique objects, and that when it happens to return a preexisting object somewhere that this throws things off. Will have to investigate further. 

### Version

1.1.6

### Component

Other

### What browsers are you seeing the problem on?

None

### Relevant log output

```shell
======================================================================
FAIL: test_project_list_search (coldfront.core.project.test_views.ProjectListViewTest.test_project_list_search)
Test that project list search works.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/eric/Work/coldfront/coldfront/core/project/test_views.py", line 304, in test_project_list_search
    self.assertEqual(len(response.context["object_list"]), 1)
AssertionError: 2 != 1

----------------------------------------------------------------------
```

### Tasks/ user tests when bug is fixed

- [ ]

--- a/coldfront/core/allocation/tests/test_models.py
+++ b/coldfront/core/allocation/tests/test_models.py
@@ -61,7 +61,7 @@ class AllocationModelCleanMethodTests(TestCase):
 
     def test_status_is_expired_and_end_date_not_past_has_validation_error(self):
         """Test that an allocation with status 'expired' and end date in the future raises a validation error."""
-        end_date_in_the_future: datetime.date = (timezone.now() + datetime.timedelta(days=1)).date()
+        end_date_in_the_future: datetime.date = datetime.date.today() + datetime.timedelta(days=1)
         actual_allocation: Allocation = AllocationFactory.build(
             status=self.expired_status, end_date=end_date_in_the_future, project=self.project
         )
@@ -70,7 +70,7 @@ class AllocationModelCleanMethodTests(TestCase):
 
     def test_status_is_expired_and_start_date_after_end_date_has_validation_error(self):
         """Test that an allocation with status 'expired' and start date after end date raises a validation error."""
-        end_date: datetime.date = (timezone.now() + datetime.timedelta(days=1)).date()
+        end_date: datetime.date = datetime.date.today() + datetime.timedelta(days=1)
         start_date_after_end_date: datetime.date = end_date + datetime.timedelta(days=1)
 
         actual_allocation: Allocation = AllocationFactory.build(
@@ -116,7 +116,7 @@ class AllocationModelCleanMethodTests(TestCase):
 
     def test_status_is_active_and_start_date_after_end_date_has_validation_error(self):
         """Test that an allocation with status 'active' and start date after end date raises a validation error."""
-        end_date: datetime.date = (timezone.now() + datetime.timedelta(days=1)).date()
+        end_date: datetime.date = datetime.date.today() + datetime.timedelta(days=1)
         start_date_after_end_date: datetime.date = end_date + datetime.timedelta(days=1)
 
         actual_allocation: Allocation = AllocationFactory.build(
@@ -237,32 +237,32 @@ class AllocationModelExpiresInTests(TestCase):
 
     def test_end_date_is_today_returns_zero(self):
         """Test that the expires_in method returns 0 when the end date is today."""
-        allocation: Allocation = AllocationFactory(end_date=timezone.now().date())
+        allocation: Allocation = AllocationFactory(end_date=datetime.date.today())
         self.assertEqual(allocation.expires_in, 0)
 
     def test_end_date_tomorrow_returns_one(self):
         """Test that the expires_in method returns 1 when the end date is tomorrow."""
-        tomorrow: datetime.date = (timezone.now() + datetime.timedelta(days=1)).date()
+        tomorrow: datetime.date = datetime.date.today() + datetime.timedelta(days=1)
         allocation: Allocation = AllocationFactory(end_date=tomorrow)
         self.assertEqual(allocation.expires_in, 1)
 
     def test_end_date_yesterday_returns_negative_one(self):
         """Test that the expires_in method returns -1 when the end date is yesterday."""
-        yesterday: datetime.date = (timezone.now() - datetime.timedelta(days=1)).date()
+        yesterday: datetime.date = datetime.date.today() - datetime.timedelta(days=1)
         allocation: Allocation = AllocationFactory(end_date=yesterday)
         self.assertEqual(allocation.expires_in, -1)
 
     def test_end_date_one_week_ago_returns_negative_seven(self):
         """Test that the expires_in method returns -7 when the end date is one week ago."""
         days_in_a_week: int = 7
-        one_week_ago: datetime.date = (timezone.now() - datetime.timedelta(days=days_in_a_week)).date()
+        one_week_ago: datetime.date = datetime.date.today() - datetime.timedelta(days=days_in_a_week)
         allocation: Allocation = AllocationFactory(end_date=one_week_ago)
         self.assertEqual(allocation.expires_in, -days_in_a_week)
 
     def test_end_date_in_one_week_returns_seven(self):
         """Test that the expires_in method returns 7 when the end date is in one week."""
         days_in_a_week: int = 7
-        one_week_from_now: datetime.date = (timezone.now() + datetime.timedelta(days=days_in_a_week)).date()
+        one_week_from_now: datetime.date = datetime.date.today() + datetime.timedelta(days=days_in_a_week)
         allocation: Allocation = AllocationFactory(end_date=one_week_from_now)
         self.assertEqual(allocation.expires_in, days_in_a_week)
 

--- a/coldfront/core/allocation/tests/test_views.py
+++ b/coldfront/core/allocation/tests/test_views.py
@@ -165,7 +165,7 @@ class AllocationChangeViewTest(AllocationViewBaseTest):
             "attributeform-TOTAL_FORMS": "1",
             "end_date_extension": 0,
         }
-        self.url = "/allocation/1/change-request"
+        self.url = f"/allocation/{self.allocation.pk}/change-request"
 
     def test_allocationchangeview_access(self):
         """Test get request"""
@@ -178,7 +178,7 @@ class AllocationChangeViewTest(AllocationViewBaseTest):
 
         self.post_data["end_date_extension"] = 90
         self.assertEqual(len(AllocationChangeRequest.objects.all()), 0)
-        response = self.client.post("/allocation/1/change-request", data=self.post_data, follow=True)
+        response = self.client.post(self.url, data=self.post_data, follow=True)
         self.assertEqual(response.status_code, 200)
         self.assertContains(response, "Allocation change request successfully submitted.")
         self.assertEqual(len(AllocationChangeRequest.objects.all()), 1)
@@ -188,7 +188,7 @@ class AllocationChangeViewTest(AllocationViewBaseTest):
 
         self.assertEqual(len(AllocationChangeRequest.objects.all()), 0)
 
-        response = self.client.post("/allocation/1/change-request", data=self.post_data, follow=True)
+        response = self.client.post(self.url, data=self.post_data, follow=True)
         self.assertEqual(response.status_code, 200)
         self.assertContains(response, "You must request a change")
         self.assertEqual(len(AllocationChangeRequest.objects.all()), 0)

--- a/coldfront/core/project/forms.py
+++ b/coldfront/core/project/forms.py
@@ -19,10 +19,12 @@ EMAIL_DIRECTOR_EMAIL_ADDRESS = import_from_settings("EMAIL_DIRECTOR_EMAIL_ADDRES
 class ProjectSearchForm(forms.Form):
     """Search form for the Project list page."""
 
+    TITLE = "Title"
     LAST_NAME = "Last Name"
     USERNAME = "Username"
     FIELD_OF_SCIENCE = "Field of Science"
 
+    title = forms.CharField(label=TITLE, max_length=255, required=False)
     last_name = forms.CharField(label=LAST_NAME, max_length=100, required=False)
     username = forms.CharField(label=USERNAME, max_length=100, required=False)
     field_of_science = forms.CharField(label=FIELD_OF_SCIENCE, max_length=100, required=False)

--- a/coldfront/core/project/tests/test_views.py
+++ b/coldfront/core/project/tests/test_views.py
@@ -32,12 +32,10 @@ class ProjectViewTestBase(TestCase):
         cls.project = ProjectFactory(status=ProjectStatusChoiceFactory(name="Active"))
 
         user_role = ProjectUserRoleChoiceFactory(name="User")
-        project_user = ProjectUserFactory(project=cls.project, role=user_role)
-        cls.project_user = project_user.user
+        cls.project_user = ProjectUserFactory(project=cls.project, role=user_role)
 
         manager_role = ProjectUserRoleChoiceFactory(name="Manager")
-        pi_user = ProjectUserFactory(project=cls.project, role=manager_role, user=cls.project.pi)
-        cls.pi_user = pi_user.user
+        cls.pi_user = ProjectUserFactory(project=cls.project, role=manager_role, user=cls.project.pi)
         cls.admin_user = UserFactory(is_staff=True, is_superuser=True)
         cls.nonproject_user = UserFactory(is_staff=False, is_superuser=False)
 
@@ -69,8 +67,8 @@ class ProjectDetailViewTest(ProjectViewTestBase):
         # logged-out user gets redirected, admin can access create page
         self.project_access_tstbase(self.url)
         # pi and projectuser can access
-        utils.test_user_can_access(self, self.pi_user, self.url)
-        utils.test_user_can_access(self, self.project_user, self.url)
+        utils.test_user_can_access(self, self.pi_user.user, self.url)
+        utils.test_user_can_access(self, self.project_user.user, self.url)
         # user not belonging to project cannot access
         utils.test_user_cannot_access(self, self.nonproject_user, self.url)
 
@@ -80,10 +78,10 @@ class ProjectDetailViewTest(ProjectViewTestBase):
         response = utils.login_and_get_page(self.client, self.admin_user, self.url)
         self.assertEqual(response.context["is_allowed_to_update_project"], True)
         # pi has is_allowed_to_update_project set to True
-        response = utils.login_and_get_page(self.client, self.pi_user, self.url)
+        response = utils.login_and_get_page(self.client, self.pi_user.user, self.url)
         self.assertEqual(response.context["is_allowed_to_update_project"], True)
         # non-manager user has is_allowed_to_update_project set to False
-        response = utils.login_and_get_page(self.client, self.project_user, self.url)
+        response = utils.login_and_get_page(self.client, self.project_user.user, self.url)
         self.assertEqual(response.context["is_allowed_to_update_project"], False)
 
     def test_projectdetail_request_allocation_button_visibility(self):
@@ -92,27 +90,27 @@ class ProjectDetailViewTest(ProjectViewTestBase):
         # admin can see request allocation button
         utils.page_contains_for_user(self, self.admin_user, self.url, button_text)
         # pi can see request allocation button
-        utils.page_contains_for_user(self, self.pi_user, self.url, button_text)
+        utils.page_contains_for_user(self, self.pi_user.user, self.url, button_text)
         # non-manager user cannot see request allocation button
-        utils.page_does_not_contain_for_user(self, self.project_user, self.url, button_text)
+        utils.page_does_not_contain_for_user(self, self.project_user.user, self.url, button_text)
 
     def test_projectdetail_edituser_button_visibility(self):
         """Test visibility of projectdetail edit button across user levels"""
         # admin can see edit button
         utils.page_contains_for_user(self, self.admin_user, self.url, "fa-user-edit")
         # pi can see edit button
-        utils.page_contains_for_user(self, self.pi_user, self.url, "fa-user-edit")
+        utils.page_contains_for_user(self, self.pi_user.user, self.url, "fa-user-edit")
         # non-manager user cannot see edit button
-        utils.page_does_not_contain_for_user(self, self.project_user, self.url, "fa-user-edit")
+        utils.page_does_not_contain_for_user(self, self.project_user.user, self.url, "fa-user-edit")
 
     def test_projectdetail_addnotification_button_visibility(self):
         """Test visibility of projectdetail add notification button across user levels"""
         # admin can see add notification button
         utils.page_contains_for_user(self, self.admin_user, self.url, "Add Notification")
         # pi cannot see add notification button
-        utils.page_does_not_contain_for_user(self, self.pi_user, self.url, "Add Notification")
+        utils.page_does_not_contain_for_user(self, self.pi_user.user, self.url, "Add Notification")
         # non-manager user cannot see add notification button
-        utils.page_does_not_contain_for_user(self, self.project_user, self.url, "Add Notification")
+        utils.page_does_not_contain_for_user(self, self.project_user.user, self.url, "Add Notification")
 
 
 class ProjectCreateTest(ProjectViewTestBase):
@@ -129,8 +127,8 @@ class ProjectCreateTest(ProjectViewTestBase):
         # logged-out user gets redirected, admin can access create page
         self.project_access_tstbase(self.url)
         # pi, projectuser and nonproject user cannot access create page
-        utils.test_user_cannot_access(self, self.pi_user, self.url)
-        utils.test_user_cannot_access(self, self.project_user, self.url)
+        utils.test_user_cannot_access(self, self.pi_user.user, self.url)
+        utils.test_user_cannot_access(self, self.project_user.user, self.url)
         utils.test_user_cannot_access(self, self.nonproject_user, self.url)
 
 
@@ -150,9 +148,9 @@ class ProjectAttributeCreateTest(ProjectViewTestBase):
         # logged-out user gets redirected, admin can access create page
         self.project_access_tstbase(self.url)
         # pi can access create page
-        utils.test_user_can_access(self, self.pi_user, self.url)
+        utils.test_user_can_access(self, self.pi_user.user, self.url)
         # project user and nonproject user cannot access create page
-        utils.test_user_cannot_access(self, self.project_user, self.url)
+        utils.test_user_cannot_access(self, self.project_user.user, self.url)
         utils.test_user_cannot_access(self, self.nonproject_user, self.url)
 
     def test_project_attribute_create_post(self):
@@ -207,9 +205,9 @@ class ProjectAttributeUpdateTest(ProjectViewTestBase):
     def test_project_attribute_update_access(self):
         """Test access to project attribute update page"""
         self.project_access_tstbase(self.url)
-        utils.test_user_can_access(self, self.pi_user, self.url)
+        utils.test_user_can_access(self, self.pi_user.user, self.url)
         # project user, pi, and nonproject user cannot access update page
-        utils.test_user_cannot_access(self, self.project_user, self.url)
+        utils.test_user_cannot_access(self, self.project_user.user, self.url)
         utils.test_user_cannot_access(self, self.nonproject_user, self.url)
 
 
@@ -230,9 +228,9 @@ class ProjectAttributeDeleteTest(ProjectViewTestBase):
         # logged-out user gets redirected, admin can access delete page
         self.project_access_tstbase(self.url)
         # pi can access delete page
-        utils.test_user_can_access(self, self.pi_user, self.url)
+        utils.test_user_can_access(self, self.pi_user.user, self.url)
         # project user and nonproject user cannot access delete page
-        utils.test_user_cannot_access(self, self.project_user, self.url)
+        utils.test_user_cannot_access(self, self.project_user.user, self.url)
         utils.test_user_cannot_access(self, self.nonproject_user, self.url)
 
 
@@ -245,7 +243,8 @@ class ProjectListViewTest(ProjectViewTestBase):
         super(ProjectListViewTest, cls).setUpTestData()
         # add 100 projects to test pagination, permissions, search functionality
         additional_projects = [ProjectFactory() for i in list(range(100))]
-        cls.additional_projects = [p for p in additional_projects if p.pi.last_name != cls.project.pi.last_name]
+        # cls.additional_projects = [p for p in additional_projects if p.pi.last_name != cls.project.pi.last_name]
+        cls.additional_projects = additional_projects
         cls.url = "/project/"
 
     ### ProjectListView access tests ###
@@ -255,8 +254,8 @@ class ProjectListViewTest(ProjectViewTestBase):
         # logged-out user gets redirected, admin can access list page
         self.project_access_tstbase(self.url)
         # all other users can access list page
-        utils.test_user_can_access(self, self.pi_user, self.url)
-        utils.test_user_can_access(self, self.project_user, self.url)
+        utils.test_user_can_access(self, self.pi_user.user, self.url)
+        utils.test_user_can_access(self, self.project_user.user, self.url)
         utils.test_user_can_access(self, self.nonproject_user, self.url)
 
     ### ProjectListView display tests ###
@@ -264,12 +263,12 @@ class ProjectListViewTest(ProjectViewTestBase):
     def test_project_list_display_members(self):
         """Project list displays only projects that user is an active member of"""
         # deactivated projectuser won't see project on their page
-        response = utils.login_and_get_page(self.client, self.project_user, self.url)
+        response = utils.login_and_get_page(self.client, self.project_user.user, self.url)
         self.assertEqual(len(response.context["object_list"]), 1)
-        proj_user = self.project.projectuser_set.get(user=self.project_user)
+        proj_user = self.project.projectuser_set.get(user=self.project_user.user)
         proj_user.status, _ = ProjectUserStatusChoice.objects.get_or_create(name="Removed")
         proj_user.save()
-        response = utils.login_and_get_page(self.client, self.project_user, self.url)
+        response = utils.login_and_get_page(self.client, self.project_user.user, self.url)
         self.assertEqual(len(response.context["object_list"]), 0)
 
     def test_project_list_displayall_permission_admin(self):
@@ -281,13 +280,13 @@ class ProjectListViewTest(ProjectViewTestBase):
     def test_project_list_displayall_permission_pi(self):
         """Projectlist displayall option displays only the pi's projects to the pi"""
         url = self.url + "?show_all_projects=on"
-        response = utils.login_and_get_page(self.client, self.pi_user, url)
+        response = utils.login_and_get_page(self.client, self.pi_user.user, url)
         self.assertEqual(len(response.context["object_list"]), 1)
 
     def test_project_list_displayall_permission_project_user(self):
         """Projectlist displayall displays only projects projectuser belongs to"""
         url = self.url + "?show_all_projects=on"
-        response = utils.login_and_get_page(self.client, self.project_user, url)
+        response = utils.login_and_get_page(self.client, self.project_user.user, url)
         self.assertEqual(len(response.context["object_list"]), 1)
 
     ### ProjectListView search tests ###
@@ -295,13 +294,10 @@ class ProjectListViewTest(ProjectViewTestBase):
     def test_project_list_search(self):
         """Test that project list search works."""
         url_base = self.url + "?show_all_projects=on"
-        url = (
-            f"{url_base}&last_name={self.project.pi.last_name}"
-            + f"&field_of_science={self.project.field_of_science.description}"
-        )
+        url = f"{url_base}&title={self.project.title}"
         # search by project project_title
         response = utils.login_and_get_page(self.client, self.admin_user, url)
-        self.assertEqual(len(response.context["object_list"]), 1)
+        self.assertIn(self.project, response.context["object_list"])
 
 
 class ProjectRemoveUsersViewTest(ProjectViewTestBase):

--- a/coldfront/core/project/views.py
+++ b/coldfront/core/project/views.py
@@ -249,44 +249,24 @@ class ProjectListView(LoginRequiredMixin, ListView):
 
         project_search_form = ProjectSearchForm(self.request.GET)
 
+        projects = Project.objects.prefetch_related("pi", "field_of_science", "status")
+
         if project_search_form.is_valid():
             data = project_search_form.cleaned_data
             if data.get("show_all_projects") and (
                 self.request.user.is_superuser or self.request.user.has_perm("project.can_view_all_projects")
             ):
-                projects = (
-                    Project.objects.prefetch_related(
-                        "pi",
-                        "field_of_science",
-                        "status",
-                    )
-                    .filter(
-                        status__name__in=[
-                            "New",
-                            "Active",
-                        ]
-                    )
-                    .order_by(order_by)
-                )
+                projects = projects.filter(status__name__in=["New", "Active"])
             else:
-                projects = (
-                    Project.objects.prefetch_related(
-                        "pi",
-                        "field_of_science",
-                        "status",
-                    )
-                    .filter(
-                        Q(
-                            status__name__in=[
-                                "New",
-                                "Active",
-                            ]
-                        )
-                        & Q(projectuser__user=self.request.user)
-                        & Q(projectuser__status__name="Active")
-                    )
-                    .order_by(order_by)
+                projects = projects.filter(
+                    Q(status__name__in=["New", "Active"])
+                    & Q(projectuser__user=self.request.user)
+                    & Q(projectuser__status__name="Active")
                 )
+
+            # Last Name
+            if data.get("title"):
+                projects = projects.filter(title__icontains=data.get("title"))
 
             # Last Name
             if data.get("last_name"):
@@ -305,26 +285,13 @@ class ProjectListView(LoginRequiredMixin, ListView):
                 projects = projects.filter(field_of_science__description__icontains=data.get("field_of_science"))
 
         else:
-            projects = (
-                Project.objects.prefetch_related(
-                    "pi",
-                    "field_of_science",
-                    "status",
-                )
-                .filter(
-                    Q(
-                        status__name__in=[
-                            "New",
-                            "Active",
-                        ]
-                    )
-                    & Q(projectuser__user=self.request.user)
-                    & Q(projectuser__status__name="Active")
-                )
-                .order_by(order_by)
+            projects = projects.filter(
+                Q(status__name__in=["New", "Active"])
+                & Q(projectuser__user=self.request.user)
+                & Q(projectuser__status__name="Active")
             )
 
-        return projects.distinct()
+        return projects.order_by(order_by).distinct()
 
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
@@ -1088,21 +1055,18 @@ class ProjectUserDetail(LoginRequiredMixin, UserPassesTestMixin, TemplateView):
 
     def get(self, request, *args, **kwargs):
         project_obj = get_object_or_404(Project, pk=self.kwargs.get("pk"))
-        project_user_pk = self.kwargs.get("project_user_pk")
+        project_user_obj = get_object_or_404(ProjectUser, pk=self.kwargs.get("project_user_pk"))
 
-        if project_obj.projectuser_set.filter(pk=project_user_pk).exists():
-            project_user_obj = project_obj.projectuser_set.get(pk=project_user_pk)
+        project_user_update_form = ProjectUserUpdateForm(
+            initial={"role": project_user_obj.role, "enable_notifications": project_user_obj.enable_notifications}
+        )
 
-            project_user_update_form = ProjectUserUpdateForm(
-                initial={"role": project_user_obj.role, "enable_notifications": project_user_obj.enable_notifications}
-            )
+        context = {}
+        context["project_obj"] = project_obj
+        context["project_user_update_form"] = project_user_update_form
+        context["project_user_obj"] = project_user_obj
 
-            context = {}
-            context["project_obj"] = project_obj
-            context["project_user_update_form"] = project_user_update_form
-            context["project_user_obj"] = project_user_obj
-
-            return render(request, self.template_name, context)
+        return render(request, self.template_name, context)
 
     def post(self, request, *args, **kwargs):
         project_obj = get_object_or_404(Project, pk=self.kwargs.get("pk"))

